### DISCUSSION TITLE
How to make 'f16' propulsion system valid in simulation.

### DISCUSSION THREAD

**Discussion by Triwd** [JSBSim-Team/jsbsim#1067] [Q&A]
Hello developers!
    I'm using JSBsim to write a program that simulates F16 flight（use model named ‘f16’）. In the simulation, I tested that the functions of sending commands (a, e, r, t) and obtaining flight status (longitude, latitude, altitude, speed, etc.) were normal, but the propulsion system (engine) did not work. 
    To start the propulsion system I did this, which is invalid：
```python
# fdm is the FGFDMExec instance
fdm.set_property_value('fcs/mixture-cmd-norm', 0.87)
fdm.set_property_value('propulsion/magneto_cmd', 1)
fdm.set_property_value('propulsion/starter_cmd', 1)
propulsion = fdm.get_propulsion() 
n = propulsion.get_num_engines()
for j in range(n):
    propulsion.get_engine(j).init_running()
propulsion.get_steady_state()
```
   Although I effectivelly set the throttle command, the altitude keeps decreasing and the value of this properties are always zero during simulation:
- 'propulsion/total-fuel-lbs'
- 'propulsion/engine/thrust-lbs'
- 'propulsion/engine/fuel-used-lbs'
- 'propulsion/engine/pn[0]/speed'
- 'propulsion/engine/pn[0]/thrust-lbs'
- 'propulsion/engine/pn[0]/fuel-flow-lbs_sec'
- 'propulsion/engine/set-running'
- 'propulsion/engine/stalled'

I must be misunderstanding something, can someone tell me how I can make the propulsion system work properly?

**Comment by nmfarras** [comment]
>...
> Although I effectivelly set the throttle command, the altitude keeps decreasing and the value of this properties are always zero during simulation:
> 
> * 'propulsion/total-fuel-lbs'
> * 'propulsion/engine/thrust-lbs'
> * 'propulsion/engine/fuel-used-lbs'
> * 'propulsion/engine/pn[0]/speed'
> * 'propulsion/engine/pn[0]/thrust-lbs'
> * 'propulsion/engine/pn[0]/fuel-flow-lbs_sec'
> * 'propulsion/engine/set-running'
> * 'propulsion/engine/stalled'
> 
> I must be misunderstanding something, can someone tell me how I can make the propulsion system work properly?

I would suggest you to try plot total fuel through out simulation. if you see the fuel keep decreasing and not what you think it supposed to be, try adjusting following

On f16.xml file:
```xml
  <tank type="FUEL"> <!-- Tank number 0 -->
   <location unit="IN">
    <x> -174.4 </x>
    <y> 65.0 </y>
    <z> 5.0 </z>
   </location>
   <capacity unit="LBS"> 3486 </capacity>
   <contents unit="LBS"> 1500 </contents>
  </tank>
```
I think you might consider adjusting the contents for all fuel tank for longer flight time. I don't know if it is what you looking for.

**Comment by seanmcleod** [comment]
Mixture and magneto are only relevant to piston engines, not jet/turbine engines.

I use the following in my python only scripts.

```python
fdm.load_model('737') 

# Set engines running
fdm['propulsion/engine[0]/set-running'] = 1
fdm['propulsion/engine[1]/set-running'] = 1
```

**Reply by Triwd** [reply]
Thanks for your answer, f16 only has one engine, so I used` fdm['propulsion/engine[0]/set-running'] = 1,` but according to the simulation process data, there is still no engine activated. In order to rule out other factors in the project, I am going to write a test script to test these properties, thank you again for your answer.

**Comment by Triwd** [comment]
I'm very sorry for taking up everyone's time. I have eliminated the problem. The reason is that I encapsulated the simulation function in the gym environment and called `fdm.reset_to_initial_conditions(0)` during env.reset to overwrite `fdm['propulsion/engine/set-running'] = 1`.
Thank you very much for every answer, "Mixture and magneto are only relevant to piston engines, not jet/turbine engines." is correct, also use `fdm['propulsion/engine/set-running'] = 1` to start the engine.

**Comment by seanmcleod** [comment]
No problem. And thanks for coming back with a detailed answer. Now someone in the future who does an internet search for something like "JSBSim gym engine won't start F-16 ..." has a chance of finding your answer. Compared to some users who simply post a question and figure out the issue, possibly after a couple of suggestions, but simply go quiet without reporting back on what actually solved the issue.
